### DISCUSSION TITLE
$$Undrained_Poisson_Ratio flag not created in GOM file for Saturated Undrained Effective Stress Analyses

### DISCUSSION THREAD

**Discussion by WaveHello** [Anura3D/Anura3D_OpenSource#131] [General]
Hello All,

When using the current release of Anura3D_2023, and running a Saturated Undrained Effective Stress Analysis there is nowhere to enter a value for the Undrained Poisson Ratio (nu_u) in GiD and the $$Undrained_Poisson_Ratio flag is not created in the GOM file. This creates an error when the Material Data is read in by the ReadMaterialData.For file for the following constitutive relations:

1) ESM_Linear_Elasticity
2) ESM_MOHR_Coloumb_Standard
3) ESM_MOHR_COULOMB_TEUNISSEN
4) ESM_MOHR_COULOMB_STRAIN_SOFTENING
5) ESM_MODIFIED_CAM_CLAY
 
When those constitutive relations are used they make a call to read nu_u and since the flag isn't created it causes the read in to be off by a row/Material property (eg. For MC the friction angle is read in as nu_u, then there is an error at the end of the material).

The code in the ReadMaterialData.For file that reads in nu_u for the first four constitutive models is shown below:
```fortran
if (trim(MatParams(I)%MaterialType)==SATURATED_SOIL_UNDRAINED_EFFECTIVE) then
                    call ReadReal(FileUnit, BName, DumR) ! Poisson ratio undrained, nu_undr, $$POISSON_RATIO_UNDRAINED
                    MatParams(I)%UndrainedPoissonRatio = DumR
                    n = MatParams(I)%InitialPorosity
                    nu_u = MatParams(I)%UndrainedPoissonRatio
                    nu_eff = MatParams(I)%PoissonRatio
                    E = MatParams(I)%YoungModulus
                    MatParams(I)%BulkModulusLiquid = n*(nu_u-nu_eff)*E/   &
                        ((1-2.*nu_u)*(1+nu_eff)*(1-2.*nu_eff))
                  else 
                    MatParams(I)%UndrainedPoissonRatio = 0.495 
                  end if                      
```
For ESM_MODIFIED_CAM_CLAY the call is slightly modified to be:
``` fortran
if (trim(MatParams(I)%MaterialType)==SATURATED_SOIL_UNDRAINED_EFFECTIVE) then
                    call ReadReal(FileUnit, BName, DumR) ! Poisson ratio undrained, nu_undr, $$POISSON_RATIO_UNDRAINED
                    MatParams(I)%UndrainedPoissonRatio = DumR
                    ! bulk modulus is intialized elsewhere because effective stress is needed
                  else
                    MatParams(I)%UndrainedPoissonRatio = 0.495 
                  end if
```

It's possible to manually create the $$Undrained_Poisson_Ratio flag in the GOM file. When using the MC_standard constitutive model, $$Undrained_Poisson_Ratio flag and correponding value should be placed right above the friction angle flag. For the other constitutive models the location varies slightly.

Thanks!

**Comment by Marlon-li** [comment]
Dear WaveHello，
   By looking at your account, I understand that you have a very deep research base on ESM, I currently have an ontology model umat, but there are some problems in converting it to an ESM.dll file, and I would be very happy to share this ontology model with you. I am very happy to share this native model with you and look forward to your help. Here is my email: 1042262646@qq.com. Looking forward to your reply.

**Comment by WaveHello** [comment]
Hello Marlon-li,

I've had similar trouble compiling ESMs as .dlls. To get around the problem it's possible to directly hardcode the UMAT into the Anura3D source code. 
[Here's a link to instructions to go about that](https://github.com/WaveHello/Anura3D-Info/blob/main/Fortran_Info/Hard_Coding_ESM_Guide.md).

Let me know if you have any questions.

**Reply by Marlon-li** [reply]
Dear WaveHello,
Thank you very much for your response. I have a clay hypoplastic constitutive model (UMAT) that runs perfectly in ABAQUS, and I'm looking to integrate it into Anura3D. As you mentioned, I encountered some issues when generating the DLL file. A few days ago, I came across your tutorial on hardcoding the UMAT into the Anura3D source code, and I got really excited because this is crucial for me. However, after trying various methods, I still face errors related to scalar or array declarations when trying to build the solution in Visual Studio.
Thank you for your assistance.

**Reply by WaveHello** [reply]
Hello Marlon-li,

To make this conversation easier to find in the future, I created a [new discussion post](https://github.com/Anura3D/Anura3D_OpenSource/discussions/181). I provided some more information there. 

Without knowing which scalars and arrays are not being compiled correctly I don't have a way of pinpointing your problem. Please provide more information about the compiler errors on the new discussion post. 

If you haven't tried stepping through the code with the Visual Studio debugger, I'd recommend that. For information on debugging in Visual Studio see the **Debugging Procedure** section [here](https://github.com/WaveHello/Anura3D-Info/blob/main/Fortran_Info/Anura3D_Help_Guide.md) or google Visual Studio debugging.

**Reply by Marlon-li** [reply]
Dear WaveHello,
Thank you very much for your help. Your patience and selflessness will help more people.

**Comment by Marlon-li** [comment]
Dear WaveHello，
 I'm very sorry to bother you again. I apologize for my lack of programming skills. My friend and I have been trying to hardcode our intrinsic model into the anura3d software for the past week, but we always get a lot of reported errors all of a sudden after they have been reduced because we don't have enough programming foundation. Since I'm facing a Ph.D. graduation in May 2024, I'm still lacking simulation chapters. I don't necessarily need the advanced constitutive models, so I am taking the liberty of asking you if you have hard-coded any constitutive models related to the overconsolidated clay soil, including the modified Cam-clay model. And whether you can share the program. I would be grateful.
  Looking forward to your reply.

**Reply by WaveHello** [reply]
Please see my response [here](https://github.com/Anura3D/Anura3D_OpenSource/discussions/181)

**Comment by GBarbosaS** [comment]
"Dear colleagues, in ANURA3D 2024 opened by Gid, we have options for constitutive models such as Rigid Body, Linear Elastic, Mohr-Coulomb, and external material model. I have seen many articles using other models, and I mainly need a constitutive model with softening; in the case of Mohr-Coulomb, softening would suffice. Can I find a DLL file with a softening model already implemented, or do I need to implement a new model myself?"

**Reply by WaveHello** [reply]
Hello GBarbosaS,

There isn't a .dll but there is an implemented MCSS model hosted [here](https://github.com/CriticalSoilModels/MohrCoulomb_StrainSoft/tree/main) . Currently it's set up for use with incrementalDriver. The function reorder_real_array and the associated function calls can be removed when the MCSS model is used in Anura3D.
